src/utils/include/c_resource.hpp

```diff
@@ -0,0 +1,366 @@
+        if constexpr (std::is_pointer_v<T>) {
+            _element = initValue;
+            constructFunction(&_element, std::forward<Ts>(Args)...);
+        } else {
+            _element = [](auto &&...args) -> decltype(auto) { // construct wrapper to allow for default constructor arguments
+                return constructFunction(std::forward<decltype(args)>(args)...);
+            }(std::forward<Ts>(Args)...);
+        }
+    }
+
+    template<typename... Ts>
+        requires(std::is_invocable_v<TConstructor, T *, Ts...>)
+    [[nodiscard]] constexpr auto emplace(Ts &&...Args) noexcept
+        requires std::is_pointer_v<T>
+    {
+        destruct(_element);
+        _element = initValue;
+        return construct(&_element, static_cast<Ts &&>(Args)...);
+    }
+
+    [[nodiscard]] constexpr c_resource(c_resource &&other) noexcept
+        requires std::is_pointer_v<T>
+    {
+        _element       = other._element;
+        other._element = initValue;
+    };
+    constexpr c_resource &operator=(c_resource &&rhs) noexcept
+        requires std::is_pointer_v<T>
+    {
+        if (this != &rhs) {
+            destruct(_element);
+            _element     = rhs._element;
+            rhs._element = initValue;
+        }
+        return *this;
+    };
+    constexpr void swap(c_resource &other) noexcept {
+        auto ptr       = _element;
+        _element       = other._element;
+        other._element = ptr;
+    }
+
+    constexpr ~c_resource() noexcept = delete;
+    constexpr ~c_resource() noexcept
+        requires destructible && std::is_pointer_v<T>
+    {
+        destruct(_element);
+    }
+    constexpr ~c_resource() noexcept
+        requires(destructible && !std::is_pointer_v<T>)
+    {
+        if constexpr (unconditionallyDestruct) {
+            destructFunction();
+        } else {
+            if (_element) {
+                destructFunction();
+            }
+        }
+    }
+
+    constexpr void clear() noexcept
+        requires destructible
+    {
+        if constexpr (std::is_pointer_v<T>) {
+            destruct(_element);
+            _element = initValue;
+        } else {
+            destructFunction();
+        }
+    }
+    constexpr c_resource &operator=(std::nullptr_t) noexcept {
+        clear();
+        return *this;
+    }
+
+    [[nodiscard]] constexpr explicit operator bool() const & noexcept {
+        return _element != initValue;
+    }
+    [[nodiscard]] constexpr explicit operator bool() && { // N.B. suppresses user-code error
+        static_assert(detail::always_false<T>, "\n**** you probably did not intend this. The resource must be assigned to an lvalue! ****\n"
+                                               "**** e.g. potential error: if (ImScoped::Window(...)) // as the temporary would be destroyed immediately ****");
+        return false;
+    }
+
+    [[nodiscard]] constexpr bool        empty() const noexcept { return _element == initValue; }
```

Comment:
Maybe _element should have been std::optional?

Reply:
The `_element` needs to be accessible because it can also carry a user-overridable default value. The default return signature of `empty()` is typically bool.

Reply:
Return type bool is fine. Not sure I understand the benefit of user-overridable default value for an empty state compared to std::optional. Without optional, it requires that `decltype(_element)` has a special value that will be used as an empty state.